Admin feature

sarad-singh opened the pull request:
- admin signin, logout
- create employee
- delete log
- give feedback(comment) to employee log
- view redesigns

Commit:
wip

Commit:
add admin routes, controllers & views

Commit:
add admin signin

Commit:
add admin dashboard

Commit:
add logout feature

Commit:
add log deletion feature

Commit:
admin feature, views redesign

Commit:
use session, flash messages in employee

Commit:
user session and flash messages for admin

Commit:
cleanup & add createEmployee for admin

Commit:
admin nav changed

Commit:
update

Commit:
add database migration

Commit:
add scrips & remove unused dependencies

Commit:
fix typo

Commit:
change index page content

Commit:
remove semicolons

Commit:
fix table field name

Commit:
register capitalize method for hbs

chandan-introcept commented on src/database/migrate.ts:
```diff
@@ -0,0 +1,32 @@
+import { db } from "./db";
+
+const allQueries = async (): Promise<void> => {
+    const employeeTable = `CREATE TABLE employee ( id INT NOT NULL AUTO_INCREMENT , name VARCHAR(255) NOT NULL , department VARCHAR(255) NOT NULL , email VARCHAR(255) NOT NULL , password VARCHAR(255) NOT NULL, isAdmin TINYINT(1), PRIMARY KEY (id), UNIQUE email (email))`
+    const logTable = `CREATE TABLE log ( id INT NOT NULL AUTO_INCREMENT , title VARCHAR(255) NOT NULL , description TEXT NOT NULL , createdDate DATETIME NOT NULL, employeeId INT NOT NULL , PRIMARY KEY (id), INDEX emplyeeId (employeeId), FOREIGN KEY (employeeId) REFERENCES employee (id))`
+    const commentTable = `CREATE TABLE comment ( id INT NOT NULL AUTO_INCREMENT , comment TEXT NOT NULL , commentedOn DATETIME NOT NULL, commentedBy INT NOT NULL , logId INT NOT NULL, PRIMARY KEY (id), INDEX commentedBy (commentedBy), INDEX logId (logId), FOREIGN KEY (commentedBy) REFERENCES employee (id), FOREIGN KEY (logId) REFERENCES log (id))`
```
please handle Referential Integrity.
you can take reference from this link
https://www.javatpoint.com/mysql-on-delete-cascade

chandan-introcept commented on src/middlewares/validations/admin/admin.ts:
```diff
@@ -0,0 +1,74 @@
+    if (!createEmployee.email) {
+        errors.email = 'email is required'
+    }
+    if (!createEmployee.email.includes('@')) {
+        errors.email = 'email must be valid email'
```
please validate email properly..we can use regex for this

chandan-introcept commented on src/models/comment.ts:
```diff
@@ -0,0 +1,66 @@
+    where ?
+    ORDER BY commentedOn DESC`
+
+    const result = await db.query(query, [{ logId }])
```
can be comments for readablity..please do proper naming of variables

chandan-introcept commented on src/models/log.ts:
```diff
@@ -1,38 +1,82 @@
-    const query = "SELECT * FROM `LOG` WHERE ?"
-    const result: Log[] = await db.query(query, [param])
-    return result
+const find = async (param?: { key: 'id' | 'employeeId', value: number }): Promise<Log[]> => {
```
we are finding employee by employee id and employee id is unique so, how can results will be multiple

chandan-introcept commented on src/services/admin.ts:
```diff
@@ -0,0 +1,42 @@
+const createComment = async (comment: string, commentedBy: number, logId: number): Promise<boolean> => {
+    const createComment: CreateComment = {
+        comment,
+        commentedOn: new Date(),
```
I think new Date() will return value in `Mon Sep 20 2021 13:34:19 GMT+0545 (Nepal Time)` please format it

chandan-introcept commented:
PR is too large for review. From next time please open PR for single features !!

Commit:
resolve pr #7
- validate for unique employee email
- referential integrity
- validate email with regular expression

Commit:
update readme

Commit:
remove semicolons

Commit:
add password hashing

Commit:
move @types/bcrypt to dev-dependencies

Commit:
check email pattern for signin

Commit:
take admin ceredentials from environment variables while  migrating

Commit:
see employees feature in admin portal

Commit:
add delete employee feature
